hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/UpdateHoodieTableCommand.scala

```diff
@@ -0,0 +1,115 @@
+import org.apache.spark.sql.catalyst.plans.logical.{Assignment, SubqueryAlias, UpdateTable}
+import org.apache.spark.sql.execution.command.RunnableCommand
+import org.apache.spark.sql.hudi.HoodieOptionConfig
+import org.apache.spark.sql.hudi.HoodieSqlUtils._
+import org.apache.spark.sql.types.StructField
+
+import scala.collection.JavaConverters._
+
+case class UpdateHoodieTableCommand(updateTable: UpdateTable) extends RunnableCommand {
+
+  private val table = updateTable.table
+  private val tableAlias = table match {
+    case SubqueryAlias(name, _) => name
+    case _ => throw new IllegalArgumentException(s"Illegal table: $table")
+  }
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    logInfo(s"start execute update command for $tableAlias")
+    def cast(exp:Expression, field: StructField): Expression = {
+      castIfNeeded(exp, field.dataType, sparkSession.sqlContext.conf)
+    }
+    val name2UpdateValue = updateTable.assignments.map {
+      case Assignment(attr: AttributeReference, value) =>
+        attr.name -> value
+    }.toMap
+
+    val updateExpressions = table.output
+      .map(attr => name2UpdateValue.getOrElse(attr.name, attr))
+      .filter { // filter the meta columns
+        case attr: AttributeReference =>
+          !HoodieRecord.HOODIE_META_COLUMNS.asScala.toSet.contains(attr.name)
+        case _=> true
+      }
+
+    val projects = updateExpressions.zip(removeMetaFields(table.schema).fields).map {
+      case (attr: AttributeReference, field) =>
+        Column(cast(attr, field))
+      case (exp, field) =>
+        Column(Alias(cast(exp, field), field.name)())
+    }
+
+    var df = Dataset.ofRows(sparkSession, table)
+    if (updateTable.condition.isDefined) {
+      df = df.filter(Column(updateTable.condition.get))
+    }
+    df = df.select(projects: _*)
+    val config = buildHoodieConfig(sparkSession)
+    df.write
+      .format("hudi")
+      .mode(SaveMode.Append)
+      .options(config)
+      .save()
+    table.refresh()
+    logInfo(s"finish execute update command for $tableAlias")
+    Seq.empty[Row]
+  }
+
+  private def buildHoodieConfig(sparkSession: SparkSession): Map[String, String] = {
+    val targetTable = sparkSession.sessionState.catalog
+      .getTableMetadata(TableIdentifier(tableAlias.identifier, tableAlias.database))
+    val path = getTableLocation(targetTable, sparkSession)
+      .getOrElse(s"missing location for $tableAlias")
+
+    val primaryColumns = HoodieOptionConfig.getPrimaryColumns(targetTable.storage.properties)
+
+    assert(primaryColumns.nonEmpty,
+      s"There are no primary key in table $tableAlias, cannot execute update operator")
+    withSparkConf(sparkSession, targetTable.storage.properties) {
+      Map(
+        "path" -> removeStarFromPath(path.toString),
+        RECORDKEY_FIELD_OPT_KEY -> primaryColumns.mkString(","),
+        KEYGENERATOR_CLASS_OPT_KEY -> classOf[ComplexKeyGenerator].getCanonicalName,
```

Comment:
The `ComplexKeyGenerator` is the default KeyGenerator for sql, because sql has the primary keys definition  in the DDL. 
User can also define custom KeyGenerator when create table, just like this:
create table h0 (id int,name string) using hudi options(
  hoodie.datasource.write.keygenerator.class = 'xxx.MyKeyGenerator'
)